sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/arithmetic.scala

```diff
@@ -103,63 +91,66 @@ private[sql] object BinaryArithmetic {
   def unapply(e: BinaryArithmetic): Option[(Expression, Expression)] = Some((e.left, e.right))
 }
 
-case class Add(left: Expression, right: Expression) extends BinaryArithmetic {
+case class Add(left: Expression, right: Expression)
+  extends BinaryArithmetic with ExpectsInputTypes {
+
   override def symbol: String = "+"
   override def decimalMethod: String = "$plus"
 
+  override def inputTypes: Seq[AbstractDataType] = Seq(NumericType, NumericType)
```

Comment:
Logically we only need to give one expect type as `BinaryOperator`s have 2 same types. We can improve this in follow-up PR.